.github/workflows/static-checks.yml

```diff
@@ -51,12 +51,33 @@ jobs:
         with:
           distribution: 'zulu'
           java-version: ${{ matrix.java }}
-          cache: 'maven'
+
+      - name: restore maven dependencies
+      # restore maven dependencies only. 'depsonly' keys are safe to restore
+      # since they don't contain any artifacts created by mvn install
+      - uses: actions/cache/restore@v3
+        with:
+          key: maven-${{ runner.os }}-depsonly-${{ hashFiles('**/pom.xml') }}
+          path: ~/.m2/repository
+          restore-keys: |
+            maven-${{ runner.os }}-depsonly-
```

Comment:
I think this should be removed as it could result in dragging outdate artifacts forward:
* project declares `artifact#1.0`
* cache stores `artifact#1.0`
* pom changes to use `artifact#2.0`
* because of fallback next cache will still contain `artifact#1.0`

Reply:
I'm following the approach suggested in https://github.com/actions/cache/blob/main/examples.md#java---maven.

Even if this includes older artifacts I believe it would still save us time compared to downloading all artifacts again on every change to them pom. This is no different than what happens when developing locally. I rarely have to clean my local .m2 cache, and it does not grow that fast.

GitHub also limits the size of cached artifacts, so I expect the size to stay bounded. If this becomes a problem we can revisit.

Reply:
if we add a `mvn dependency:purge-local-repository` step before compile, would that make you feel better?

Reply:
I'm not sure how `purge-local-repository` would help here ; wouldn't that still remove the local repo in case there is a cache hit?

I wonder if the problem is with the opportunistic caching of project artifacts; can't we somehow avoid them getting into the cache:
* exclude them from the cache
   * I did experiment with this - and its a bit convoluted due to the fact that excludes work by filtering the included list [see here](https://github.com/actions/cache/issues/364)
* remove them with an `always()` [step](https://github.com/kgyrtkirk/druid/blob/a50ceea91f93e43f940834675bc53a19b9f50f6b/.github/workflows/reusable-unit-tests.yml#L71C1-L74C51)

I think this second approach is better - because if someone decides to add another exclude - it will be less likely to trigger the exclusion issue differently

Reply:
- we can use `purge-local-repository` to do what you suggest, it has lots of options to include/exclude various artifacts / purge only snapshot, and doesn't make any assumptions where the maven repo is stored.
- I fail to see how your suggestion addresses your earlier point that the cache would just keep growing though.